### PR TITLE
added `preg_filter` return-type extension

### PR DESCRIPTION
**Pull request by staabm**
closes https://github.com/phpstan/phpstan/issues/6664

### CODE REVIEW

File: src/Type/Php/ArrayRandFunctionReturnTypeExtension.php [hunk 1]
```diff
@@ -27,7 +27,7 @@ public function isFunctionSupported(FunctionReflection $functionReflection): boo
 	public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): Type
 	{
 		$argsCount = count($functionCall->getArgs());
-		if (count($functionCall->getArgs()) < 1) {
+		if ($argsCount < 1) {
```
Comment (staabm) [inline src/Type/Php/ArrayRandFunctionReturnTypeExtension.php:30]: drive-by cleanup

File: src/Type/Php/PregFilterFunctionReturnTypeExtension.php [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+		return $functionReflection->getName() === 'preg_filter';
+	}
+
+	public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): Type
+	{
+		$defaultReturn = ParametersAcceptorSelector::selectSingle($functionReflection->getVariants())->getReturnType();
+
+		$argsCount = count($functionCall->getArgs());
+		if ($argsCount < 3) {
+			return $defaultReturn;
```
Comment (ondrejmirtes) [inline src/Type/Php/PregFilterFunctionReturnTypeExtension.php:32]: This branch isn't tested

File: src/Type/Php/PregFilterFunctionReturnTypeExtension.php [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+		$subjectType = $scope->getType($functionCall->getArgs()[2]->value);
+
+		if ($subjectType->isArray()->yes()) {
+			return new ArrayType(new IntegerType(), new StringType());
+		}
+		if ($subjectType->isString()->yes()) {
+			return new UnionType([new StringType(), new NullType()]);
+		}
+
+		return $defaultReturn;
```
Comment (ondrejmirtes) [inline src/Type/Php/PregFilterFunctionReturnTypeExtension.php:44]: This branch isn't tested

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

**Comment by arnaud-lb** [comment]
Thank you @staabm !

**Pull request merged**